### ISSUE TITLE
DSPO Training Loss Trend Issue

### ISSUE THREAD

**Issue by lucasxu777** [huaishengzhu/DSPO#5]
Thanks for the amazing work here. 

I am just wondering if you guys seem to have this issue too? After plotting the step loss from training using Diffusion DPO's training script, I think there is no effective learning going on here; however, their winning rate is indeed increased when evaluating on 500 prompt examples from pickscorev2 dataset, matching to what they put in their paper. I am having the same issue with DSPO, which is a newer paper built upon the Diffusion DPO paper. 

<img width="986" alt="Image" src="https://github.com/user-attachments/assets/7a1ac533-e9ad-442e-accc-ab5037fd165a" />

**Comment by huaishengzhu** [comment]
Hi,

Thanks for your interest about our work! I guess that the loss behavior during SFT training may exhibit patterns similar to those observed with DPO objectives, although I didn’t record the loss curve. A similar phenomenon is often seen when training diffusion models—where the loss does not consistently decrease, even though the model’s performance continues to improve.

One possible explanation is related to the stochastic nature of timestep sampling during training. At each iteration, the model optimizes for a randomly sampled timestep. While the loss may decrease for one timestep in a given iteration, the next iteration may sample a different timestep where the loss hasn’t yet improved. As a result, the overall loss curve may appear noisy or stagnant across adjacent optimization steps when the loss curve is being made at the per-timestep level.

Best,

Huaisheng

**Comment by lucasxu777** [comment]
Thanks! But what's really counter-intuition here is that if the training loss looks like spikes shape here, how do we know if the model is learning consistently? Because the loss function is designed to help the model learn during the training sommothly (maybe with some spikes, but the overall trend should be a clear downwards curve, correct?) With the spike shape, I am not quite sure how the model is learning meaningfully during the training, yet has improvements.

**Comment by huaishengzhu** [comment]
Yes, your concern is really reasonable. From my experience, determining whether the diffusion model is learning consistently is much less straightforward only from loss curve. At this point, my current thinking is to evaluate the model's performance on a validation set at regular intervals (maybe potentially after every few iterations or even after each epoch) to observe whether the model consistently learn. Btw, have you tried aggregating the loss over multiple optimization steps (e.g., averaging the loss across 100 consecutive steps) to see whether the overall trend shows a clearer downward pattern? That's my opinion for this problem.

Best,

Huaisheng

**Comment by lucasxu777** [comment]
The only things I have done so far is using your loss function but modifying the hyperparameters for fine tuning little bit here, and then run evaluations when the checkpoints are ready on different datasets that are mentioned in your paper. What you just suggested sound like a real-time wise evaluation?

**Comment by huaishengzhu** [comment]
For practical implementation, one approach is to save checkpoints every 50 or 100 iterations and evaluate them on the validation set. This allows you to select the best-performing checkpoint based on actual performance, rather than relying solely on training loss.

**Comment by lucasxu777** [comment]
I see so you are saving checkpoints every fewer steps and perform evaluation.
